### PR TITLE
Use hydra.run.dir (not os.getcwd) for DDP subprocesses' run dir

### LINKED ISSUE #16694: Change the command to initialize DDP subprocess with Hydra
### Description & Motivation

Hi, 

the current behavior to initialize DDP subprocesses with Hydra is [here](https://github.com/Lightning-AI/lightning/blob/accd2b9e61063ba3c683764043030545ed87c71f/src/lightning/fabric/strategies/launchers/subprocess_script.py#L159), which overrides the hydra.run.dir with os.getcwd(). 

This is assuming the os.getcwd() will return the output directory that Hydra assigned. However, since Hydra 1.2, Hydra will keep the cwd as the project directory by default. This command will result in logging the subprocess in the project directory. 

### Pitch

Logging the DDP subprocess correctly in the output directory created by Hydra

### Alternatives

use:
```
os_cwd = hydra.core.hydra_config.HydraConfig.get()['runtime']['output_dir']
os_cwd = f'"{os_cwd}"'

```
instead

### Additional context

_No response_

cc @borda @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by nisheethlahoti**
## What does this PR do?

Fixes #16694

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by nisheethlahoti** [comment]
Not sure if failing tests (all related to deepspeed integration) are related to this PR. I can't see hydra being used anywhere in deepspeed setup, and this PR only changes the `_hydra_subprocess_cmd` function.

**Comment by ishandutta0098** [comment]
@aniketmaurya can you look into this once? We discussed this in our call yesterday.

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
@nisheethlahoti Thanks for the PR. The change here looks good overall, but we need to make sure we don't regress on the issue here: #15689
There was a PR a while ago that change the behavior of the launcher wrt. Hydra too (#11617) but we had to revert it (see discussion here #15689). 

If you would be so kind to go over these two issues to check whether there are any conflicts between your approach and the concerns raised there. That would be very helpful (forgive me, I'm a bit unfamiliar with hydra). 

> All the failing tests are failing with "ModuleNotFound: torch", which cannot possibly be caused by this PR.

Don't worry about this, we will take care of it once the PR can be merged :)

**Comment by nisheethlahoti** [comment]
> If you would be so kind to go over these two issues to check whether there are any conflicts between your approach and the concerns raised there.

The behaviour defined in #11617 actually seems better than what's proposed here (since it handles multi-run properly and has separate subdirectories for all the different PL processes), with 1 exception: that it actually tries to read the config from the `.hydra` directory (the problem raised in #15689).  This PR doesn't have that issue since it doesn't try to read any stored config (It accesses run.dir through `HydraConfig.get()`, which is present whether or not a config is stored).

However, this could also be fixed by reinstating the code in #11617, removing the line `command += ["-cp", hydra_output, "-cn", "config.yaml"]`, and running the new commands `cwd=get_original_cwd()` (as done in #15737). This would add multirun support again, while not breaking the NeMo use-case, IMO. Should I modify this PR (or send a new one) to do that instead?

**Comment by nisheethlahoti** [comment]
Update to previous comment: Removing the `-cp hydra_output` line will actually disable multi-run support, but we will still have:
1. Creation of different subdirectories for all spawned processes.
2. Support for multi-run if user has enabled hydra's experimental-rerun callback

**Comment by awaelchli** [comment]
Thanks @nisheethlahoti for your helpful analysis! Based on this I suggest we move forward with your PR as is. 

> However, this could also be fixed by reinstating the code in https://github.com/Lightning-AI/lightning/pull/11617 [...] This would add multirun support again, while not breaking the NeMo use-case, IMO. Should I modify this PR (or send a new one) to do that instead?

This is good news. I would suggest that we break it in two parts, this PR + multi-run in a different PR.

**Comment by awaelchli** [comment]
Would it be possible to add a small unit test in `tests/tests_fabric/strategies/launchers/test_subprocess_script.py`? Happy to help out if needed.

**Comment by nisheethlahoti** [comment]
> Would it be possible to add a small unit test in `tests/tests_fabric/strategies/launchers/test_subprocess_script.py`? Happy to help out if needed.

Done

File: tests/tests_pytorch/strategies/launchers/test_subprocess_script.py
Comment (awaelchli) [inline tests/tests_pytorch/strategies/launchers/test_subprocess_script.py:63]: micro nit: we can replace the `tmpdir` fixture with `tmp_path`

File: tests/tests_pytorch/strategies/launchers/test_subprocess_script.py
Comment (awaelchli) [inline tests/tests_pytorch/strategies/launchers/test_subprocess_script.py:53]: Is "null" simulating the same we already have under the None case?

File: src/lightning/fabric/strategies/launchers/subprocess_script.py
Comment (awaelchli) [inline src/lightning/fabric/strategies/launchers/subprocess_script.py:146]: The minimun supported version according to our imports is `hydra-core>=1.0.7`, would you mind double checking that this import works for that version. Thx

File: src/lightning/fabric/strategies/launchers/subprocess_script.py
Comment (awaelchli) [inline src/lightning/fabric/strategies/launchers/subprocess_script.py:160]: The `.pl_ddp_hydra_local_rank` is a new addition here too. This is the subdir we choose if the hydra config does not provide the output dir. Is this the default behavior, correct? Perhaps we should add an additional assertion to the test that this folder got created.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/fabric/strategies/launchers/subprocess_script.py
Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:146]: nvm we import it in other places too

File: tests/tests_pytorch/strategies/launchers/test_subprocess_script.py
Comment (nisheethlahoti) [reply tests/tests_pytorch/strategies/launchers/test_subprocess_script.py:53]: No, these are different. Subdir being `"null"` leads to passing `hydra.output_subdir=null` to the subprocess, which leads to `output_subdir` being `None` inside the subprocess, which leads to there being no directory to store `config.yaml` etc. (Which is the NeMo use-case).

Subdir being `None`, on the other hand, leads to the check in line 65, according to which we don't pass any `hydra.output_subdir` term, which leads to `output_subdir` having its default value (`.hydra`) inside the subprocess.

Having written this out, it sounds pretty confusing. Not sure what we could change it to, though. Open to suggestions.

File: src/lightning/fabric/strategies/launchers/subprocess_script.py
Comment (nisheethlahoti) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:160]: `.pl_ddp_hydra_{local_rank}` is actually a subdir, not the output dir (basically, a replacement to the `.hydra` subdirectory that's created inside the output dir).

I think this change makes sense because otherwise all the different processes will try to write the `config.yaml` file inside the same `.hydra` folder, which might lead to the file being garbled (not sure if the writes by different processes are atomic).

File: tests/tests_pytorch/strategies/launchers/test_subprocess_script.py
Comment (awaelchli) [reply tests/tests_pytorch/strategies/launchers/test_subprocess_script.py:52]: Thanks for the explanation. Since the rundir="null" comes from the hydra configuration of the user, this should be fine and it is not breaking any behavior. 

Since me and others from the core team can't say much about this topic (we're not hydra users), I'm fine with this. Whatever you think is intuitive and convenient for the typical Hydra user. 

What's important for us is that we don't create undesired breaking changes for large repos like NeMo which heavily use hydra. But you already commented about this and so from my side, good to move forward with this.

File: src/lightning/fabric/strategies/launchers/subprocess_script.py
Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:160]: Only hydra writes to these special folders, right? The user wouldn't see or interact with files in there (an immense noob question sorry).

Comment (nisheethlahoti) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:160]: The purpose of the special folders is to dump the configuration used for the job, if the user wants to access them later.

Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:160]: Okay I see. To me that sounds a bit strange, I can' see why the configuration file should be different between the processes. So the user would end up with many folders (e.g. 8 folders per machine if 8 gpus are used and all have the same config file inside). That might be a bit confusing. 

My proposal:
For (global) rank > 0, set `output_subdir` to null (disabling the saving of the config files). Then, only rank 0 will write the `.hydra` folder.

Comment (nisheethlahoti) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py]: Makes sense. Done.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Awesome! Thank you @nisheethlahoti 🎉

**Pull request merged**
